docs/library_java.md

```diff
@@ -0,0 +1,178 @@
+So for example: `intArg("n")` means that request's json object should has key `n` with a value that can be converted to integer.
+There are following basic methods do define an argument (they have similar names for different types: intArg/stringArg/doubleArg/booleanArg).
+For example for integer:
+- `intArg(String name)` -  required argument by name
+- `intArg(String name, int defaultValue)` - if argument is missed, function will fallback to default value
+- `optIntArg(String name)` - function will receive `Optional<Integer>`
+- `intListArg(String name)` - function will receive `List<Integer>`
+
+Method `withArgs` accepts from 1 to 21 argument and returns `ArgsN`.
+```java
+Args1 one = withArgs(intArg("n"))
+
+Args2 two = withArgs(intArg("n"), stringArg("str"))
+
+Args3 three = withArgs(intArg("n"), stringArg"str"), booleanArg("flag"))
+```
+
+#### Contexts
+
+Next to complete Mist Function definition we should inject Spark Context.
+Mist provides managed Spark Contexts, so developer does not care about context's lifecycle and settings.
+There are following methods of `ArgsN` to define Spark Context:
+- `onSparkContext`
+- `onStreamingContext`
+
+It accepts function with `n+1` arguments where `n` is count of combined argument plus `SparkContext` at end.
+```java
+JHandle<?> fromOne = withArgs(intArg("n")).onSparkContext((n, sc) -> { ... })
+
+JHandle<?> fromTwo = withArgs(intArg("n"), stringArg("str")).onSparkContext((n, s, sc) -> { ... })
+
+JHandle<?> fromThree = withArgs(intArg("n"), stringArg("str"), booleanArg("flag"))
+    .onSparkContext((n, s, b, sc) -> { ... })
+```
+
+If your job doesn't require any arguments, there are similar methods available from `JMistFn`
+```java
+import mist.api.jdsl.*;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class NoArgsExample extends JMistFn<Integer> {
+
+    @Override
+    public JHandle<Integer> handle() {
+        return onSparkContext((sc) -> RetValues.of(42));
+    }
+}
+```
+
+#### Validation
+
+For example for calculating pi using dartboard method n should be at least positive number.
+For that purpose `JArg<A>` has special methods to validate arguments:
+- `validated(f: A => Boolean)`
+- `validated(f: A => Boolean, explanation: String)`
+
+```java
+import mist.api.jdsl.*;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class JavaPiExample extends JMistFn<Double> {
+
+    @Override
+    public JHandle<Double> handle() {
+        JArg<Integer> samples = intArg("samples").validated(s -> s > 0, "Samples must be positive");
+        return withArgs(samples).onSparkContext((n, sc) -> {
+          ...
+        });
+    }
+}
+```
+
+#### Encoding
+
+Mist should be able to return result back to the client (http request, async interfaces) and it requires
+that result should be serialized to json.For that purposes function should return an instance of `RetValue`.
+There are a lot of helper methonds in `mist.api.jdsl.RetValues`
```

Comment:
we should add here supported return types because there is no support for custom encoding definition.